File: Merge sort/mergesort.py
arr=[]           

def create_the_halfs (arr, start, end):
 
 half = []
 for i in range ( start, (end+1)):
  half.append(arr[i])
 return (half)

def merge ( arr, start, mid, end):
 
 
 lower_half=[]
 upper_half=[]
 lower_half = create_the_halfs(arr, start, mid)
 upper_half = create_the_halfs(arr, (mid+1), end)

 
 llimit = len(lower_half)
 
 ulimit = len(upper_half) 
 
 i=0
 j=0
 pos=start

 while ( (i < llimit) and (j < ulimit) ): 
  
  if lower_half[i] == upper_half[j]:
   
   arr[pos] = lower_half[i]
   i=i+1
   
   pos=pos+1
   
   arr[pos] = upper_half[j]
   
   j=j+1
  
  elif lower_half[i] < upper_half[j]:
   
   arr[pos] = lower_half[i] 
   i = i + 1
  
  else:
    
   arr[pos] = upper_half[j]
   j = j + 1
    
  pos = pos + 1 
  
 while ( i < llimit):
   
   arr[pos] = lower_half[i]
   i=i+1
   pos = pos + 1
 
 while (j < ulimit):
   arr[pos] = upper_half[j]
   j= j + 1
   pos = pos + 1
  

def mergesort (arr, start, end):
 
 if ( (end -start) < 1):
  return
 
 mid = (start + end) // 2; 

 
 mergesort (arr, start, mid)
 mergesort (arr, mid+1, end)
 merge (arr, start, mid, end)
 





def accept_user_input():
 
 global arr
 print("Enter the elements: ",end="\n")
 arr=list(input().split())      
 arr=[int(i) for i in arr]
 
def print_sorted_list():
 
 print("\nElements after sorting by Mergesort:",end="\n")
 for item in arr:
  print(item,end=' ')
 print("\n") 


accept_user_input()

mergesort (arr, 0, (len(arr)-1))

print_sorted_list() 

File: Sorting/Mergesort/mergesort.cpp
#include <iostream>
#include <vector>


void merge(std::vector<int>& darr, int start, int mid, int end)
{
	std::vector<int> lh;
	std::vector<int> hh;

	for (int i = start; i <= mid; i++)
		lh.push_back(darr[i]);

	for (int i = mid + 1; i <= end; i++)
		hh.push_back(darr[i]);

	int i = 0;
	int j = 0;
	int k = start;

	while ((i < lh.size()) && (j < hh.size()))
	{
		if (lh[i] <= hh[j])
		{
			darr[k] = lh[i];
			i++;
		}
		else if (hh[j] < lh[i])
		{
			darr[k] = hh[j];
			j++;
		}

		k++;
	}

	while (i < lh.size())
	{
		darr[k] = lh[i];
		i++;
		k++;
	}

	while (j < hh.size())
	{
		darr[k] = hh[j];
		j++;
		k++;
	}
}

void mergesort(std::vector<int>& darr, int start, int end)
{
	if ((end - start) < 1)
		return;

	int mid = (start + end) / 2;

	mergesort(darr, start, mid);
	mergesort(darr, mid + 1, end);
	merge(darr, start, mid, end);

}

int main()
{
	int size;
	int element;
	std::vector<int> darray;
	std::cout << "Do enter the total count of elements:" << std::endl;
	std::cin >> size;

	std::cout << "Enter the elements: " << std::endl;
	for (int i = 0; i < size; i++)
	{
		std::cin >> element;
		darray.push_back(element);
	}

	std::cout << std::endl;

	mergesort(darray, 0, (darray.size() - 1));

	std::cout << "Elements after mergesort: " << std::endl;
	for (auto& vr: darray)
		std::cout << vr << " ";
	std::cout << std::endl;
	return 0;
}
